File: nodes/core/market/filters/widening_emas_filter_node.py
import logging

from core.types_registry import IndicatorResult, IndicatorType, IndicatorValue, OHLCVBar
from nodes.core.market.filters.base.base_indicator_filter_node import BaseIndicatorFilter
from services.indicator_calculators.ema_calculator import calculate_ema

logger = logging.getLogger(__name__)


class WideningEMAsFilter(BaseIndicatorFilter):
    """
    Filters assets based on whether the difference between two EMAs is widening or narrowing.

    Calculates the difference between two EMAs and compares it to the previous period's difference
    to determine if the EMAs are diverging (widening) or converging (narrowing).

    Formula: [EMA(fast_period) - EMA(slow_period)] compared to [previous EMA(fast_period) - EMA(slow_period)]

    Widening: Current difference > Previous difference
    Narrowing: Current difference < Previous difference

    Only assets meeting the widening/narrowing condition will pass the filter.
    """

    default_params = {
        "fast_ema_period": 10,  # Fast EMA period
        "slow_ema_period": 30,  # Slow EMA period
        "widening": True,  # True for widening, False for narrowing
    }

    params_meta = [
        {
            "name": "fast_ema_period",
            "type": "number",
            "default": 10,
            "min": 2,
            "step": 1,
            "label": "Fast EMA Period",
            "description": "Period for the faster EMA (e.g., 10)",
        },
        {
            "name": "slow_ema_period",
            "type": "number",
            "default": 30,
            "min": 2,
            "step": 1,
            "label": "Slow EMA Period",
            "description": "Period for the slower EMA (e.g., 30)",
        },
        {
            "name": "widening",
            "type": "text",
            "default": "true",
            "label": "Check for Widening",
            "description": "true: filter for widening EMAs, false: filter for narrowing EMAs",
        },
    ]

    def _validate_indicator_params(self):
        """Validate EMA period parameters."""
        fast_period = self.params.get("fast_ema_period", 10)
        slow_period = self.params.get("slow_ema_period", 30)

        if not isinstance(fast_period, (int, float)) or fast_period < 2:
            raise ValueError("Fast EMA period must be at least 2")
        if not isinstance(slow_period, (int, float)) or slow_period < 2:
            raise ValueError("Slow EMA period must be at least 2")
        if fast_period >= slow_period:
            raise ValueError("Fast EMA period must be less than slow EMA period")

    def _calculate_indicator(self, ohlcv_data: list[OHLCVBar]) -> IndicatorResult:
        """Calculate EMA difference and compare widening/narrowing."""
        if not ohlcv_data:
            return IndicatorResult(
                indicator_type=IndicatorType.EMA,
                timestamp=0,
                values=IndicatorValue(lines={"ema_difference": 0.0, "is_widening": False}),
                params=self.params,
                error="No data",
            )

        # Need enough data for both EMAs plus one comparison period
        fast_period_raw = self.params.get("fast_ema_period", 10)
        slow_period_raw = self.params.get("slow_ema_period", 30)

        if not isinstance(fast_period_raw, (int, float)):
            fast_period = 10
        else:
            fast_period = int(fast_period_raw)

        if not isinstance(slow_period_raw, (int, float)):
            slow_period = 30
        else:
            slow_period = int(slow_period_raw)

        min_data_needed = max(fast_period, slow_period) + 1

        if len(ohlcv_data) < min_data_needed:
            return IndicatorResult(
                indicator_type=IndicatorType.EMA,
                timestamp=ohlcv_data[-1]["timestamp"],
                values=IndicatorValue(lines={"ema_difference": 0.0, "is_widening": False}),
                params=self.params,
                error=f"Insufficient data: need at least {min_data_needed} bars",
            )

        # Extract close prices
        close_prices = [bar["close"] for bar in ohlcv_data]

        # Calculate EMAs using the calculator
        fast_ema_result = calculate_ema(close_prices, period=fast_period)
        slow_ema_result = calculate_ema(close_prices, period=slow_period)

        fast_ema_series = fast_ema_result.get("ema", [])
        slow_ema_series = slow_ema_result.get("ema", [])

        # Check if we have enough data for comparison
        if len(fast_ema_series) < 2 or len(slow_ema_series) < 2:
            return IndicatorResult(
                indicator_type=IndicatorType.EMA,
                timestamp=ohlcv_data[-1]["timestamp"],
                values=IndicatorValue(lines={"ema_difference": 0.0, "is_widening": False}),
                params=self.params,
                error="Unable to calculate EMA difference",
            )

        # Get the last two values from each EMA series
        current_fast = fast_ema_series[-1]
        prev_fast = fast_ema_series[-2]
        current_slow = slow_ema_series[-1]
        prev_slow = slow_ema_series[-2]

        # Check for None values (calculator returns None when insufficient data)
        if current_fast is None or prev_fast is None or current_slow is None or prev_slow is None:
            return IndicatorResult(
                indicator_type=IndicatorType.EMA,
                timestamp=ohlcv_data[-1]["timestamp"],
                values=IndicatorValue(lines={"ema_difference": 0.0, "is_widening": False}),
                params=self.params,
                error="EMA calculation returned None values",
            )

        # Calculate EMA differences for the last two periods
        current_diff = current_fast - current_slow
        prev_diff = prev_fast - prev_slow

        # Check if widening (current difference > previous difference)
        is_widening = current_diff > prev_diff

        return IndicatorResult(
            indicator_type=IndicatorType.EMA,
            timestamp=ohlcv_data[-1]["timestamp"],
            values=IndicatorValue(
                lines={
                    "ema_difference": current_diff,
                    "is_widening": is_widening,
                    "fast_ema": current_fast,
                    "slow_ema": current_slow,
                    "prev_difference": prev_diff,
                }
            ),
            params=self.params,
        )

    def _should_pass_filter(self, indicator_result: IndicatorResult) -> bool:
        """Pass filter if EMA widening/narrowing condition is met."""
        if indicator_result.error:
            return False

        lines = indicator_result.values.lines
        if "is_widening" not in lines:
            return False

        is_widening = lines["is_widening"]

        # Handle string "true"/"false" from UI
        widening_param = self.params.get("widening", True)
        if isinstance(widening_param, str):
            widening_param = widening_param.lower() == "true"

        return is_widening == widening_param

File: backend/src/nodes/core/market/filters/widening-emas-filter-node.ts
// src/nodes/core/market/filters/widening-emas-filter-node.ts
// Translated from: nodes/core/market/filters/widening_emas_filter_node.py

import { BaseIndicatorFilter } from './base/base-indicator-filter-node';
import { IndicatorType, createIndicatorResult, createIndicatorValue } from '../../../../core/types';
import type {
  ParamMeta,
  DefaultParams,
  OHLCVBar,
  IndicatorResult,
} from '../../../../core/types';
import { calculateEma } from '../../../../services/indicator-calculators';

/**
 * Filters assets based on whether the difference between two EMAs is widening or narrowing.
 *
 * Calculates the difference between two EMAs and compares it to the previous period's difference
 * to determine if the EMAs are diverging (widening) or converging (narrowing).
 *
 * Formula: [EMA(fast_period) - EMA(slow_period)] compared to [previous EMA(fast_period) - EMA(slow_period)]
 *
 * Widening: Current difference > Previous difference
 * Narrowing: Current difference < Previous difference
 *
 * Only assets meeting the widening/narrowing condition will pass the filter.
 */
export class WideningEMAsFilter extends BaseIndicatorFilter {
  static override defaultParams: DefaultParams = {
    fast_ema_period: 10,
    slow_ema_period: 30,
    widening: true,
  };

  static override paramsMeta: ParamMeta[] = [
    {
      name: 'fast_ema_period',
      type: 'number',
      default: 10,
      min: 2,
      step: 1,
      label: 'Fast EMA Period',
      description: 'Period for the faster EMA (e.g., 10)',
    },
    {
      name: 'slow_ema_period',
      type: 'number',
      default: 30,
      min: 2,
      step: 1,
      label: 'Slow EMA Period',
      description: 'Period for the slower EMA (e.g., 30)',
    },
    {
      name: 'widening',
      type: 'text',
      default: 'true',
      label: 'Check for Widening',
      description: 'true: filter for widening EMAs, false: filter for narrowing EMAs',
    },
  ];

  protected override validateIndicatorParams(): void {
    const fastPeriod = this.params.fast_ema_period ?? 10;
    const slowPeriod = this.params.slow_ema_period ?? 30;

    if (typeof fastPeriod !== 'number' || fastPeriod < 2) {
      throw new Error('Fast EMA period must be at least 2');
    }
    if (typeof slowPeriod !== 'number' || slowPeriod < 2) {
      throw new Error('Slow EMA period must be at least 2');
    }
    if (fastPeriod >= slowPeriod) {
      throw new Error('Fast EMA period must be less than slow EMA period');
    }
  }

  protected calculateIndicator(ohlcvData: OHLCVBar[]): IndicatorResult {
    if (!ohlcvData || ohlcvData.length === 0) {
      return createIndicatorResult(
        IndicatorType.EMA,
        0,
        createIndicatorValue(0, { ema_difference: 0.0, is_widening: 0 }),
        this.params,
        'No data'
      );
    }

    // Need enough data for both EMAs plus one comparison period
    let fastPeriodRaw = this.params.fast_ema_period ?? 10;
    let slowPeriodRaw = this.params.slow_ema_period ?? 30;

    const fastPeriod = typeof fastPeriodRaw === 'number' ? Math.floor(fastPeriodRaw) : 10;
    const slowPeriod = typeof slowPeriodRaw === 'number' ? Math.floor(slowPeriodRaw) : 30;

    const minDataNeeded = Math.max(fastPeriod, slowPeriod) + 1;

    if (ohlcvData.length < minDataNeeded) {
      return createIndicatorResult(
        IndicatorType.EMA,
        ohlcvData[ohlcvData.length - 1].timestamp,
        createIndicatorValue(0, { ema_difference: 0.0, is_widening: 0 }),
        this.params,
        `Insufficient data: need at least ${minDataNeeded} bars`
      );
    }

    // Extract close prices
    const closePrices = ohlcvData.map((bar) => bar.close);

    // Calculate EMAs using the calculator
    const fastEmaResult = calculateEma(closePrices, fastPeriod);
    const slowEmaResult = calculateEma(closePrices, slowPeriod);

    const fastEmaSeries = fastEmaResult.ema;
    const slowEmaSeries = slowEmaResult.ema;

    // Check if we have enough data for comparison
    if (fastEmaSeries.length < 2 || slowEmaSeries.length < 2) {
      return createIndicatorResult(
        IndicatorType.EMA,
        ohlcvData[ohlcvData.length - 1].timestamp,
        createIndicatorValue(0, { ema_difference: 0.0, is_widening: 0 }),
        this.params,
        'Unable to calculate EMA difference'
      );
    }

    // Get the last two values from each EMA series
    const currentFast = fastEmaSeries[fastEmaSeries.length - 1];
    const prevFast = fastEmaSeries[fastEmaSeries.length - 2];
    const currentSlow = slowEmaSeries[slowEmaSeries.length - 1];
    const prevSlow = slowEmaSeries[slowEmaSeries.length - 2];

    // Check for null values (calculator returns null when insufficient data)
    if (currentFast === null || prevFast === null || currentSlow === null || prevSlow === null) {
      return createIndicatorResult(
        IndicatorType.EMA,
        ohlcvData[ohlcvData.length - 1].timestamp,
        createIndicatorValue(0, { ema_difference: 0.0, is_widening: 0 }),
        this.params,
        'EMA calculation returned null values'
      );
    }

    // Calculate EMA differences for the last two periods
    const currentDiff = currentFast - currentSlow;
    const prevDiff = prevFast - prevSlow;

    // Check if widening (current difference > previous difference)
    const isWidening = currentDiff > prevDiff;

    return createIndicatorResult(
      IndicatorType.EMA,
      ohlcvData[ohlcvData.length - 1].timestamp,
      createIndicatorValue(0, {
        ema_difference: currentDiff,
        is_widening: isWidening ? 1 : 0,
        fast_ema: currentFast,
        slow_ema: currentSlow,
        prev_difference: prevDiff,
      }),
      this.params
    );
  }

  protected shouldPassFilter(indicatorResult: IndicatorResult): boolean {
    if (indicatorResult.error) {
      return false;
    }

    const lines = indicatorResult.values.lines;
    if (!('is_widening' in lines)) {
      return false;
    }

    const isWidening = lines.is_widening === 1;

    // Handle string "true"/"false" from UI
    let wideningParam = this.params.widening ?? true;
    if (typeof wideningParam === 'string') {
      wideningParam = wideningParam.toLowerCase() === 'true';
    }

    return isWidening === wideningParam;
  }
}
